File: algorithms/剑指offer/剑指 Offer 06. 从尾到头打印链表/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// 递归
class Solution {
vector<int> temp;
public:
    vector<int> reversePrint(ListNode* head) {
        if(!head) {
            return {};
        }
        if(head->next){
            reversePrint(head->next);
        }
        temp.push_back(head->val);
        return temp;
        
    }
};

File: leetcode/algorithms/0206_reverse-linked-list/java/2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
// 递归
class Solution {
    public ListNode reverseList(ListNode head) {
      // head==null 传经来的就是null
      // head.next == null已经到最后一个了
      if(head==null||head.next==null){
        return head;
      }
      // 1->2->3->4->5,返回的new_head =5
      ListNode new_head = reverseList(head.next);
      // head = 4
      head.next.next = head;
      head.next = null;
      return new_head;
    }
}